### ISSUE TITLE
Dependency (thrift) building errors on Ubuntu14.14

### ISSUE THREAD

**Issue by Red-wolf-s-husband** [facebook/osquery#974]
When I run make, I get this:
OSError: [Errno 2] No such file or directory: '/home/sunan/opensource/osquery/osquery/tables/specs/linux2'
What did I missed?

**Comment by marpaia** [comment]
are you trying to compile osquery version 0.0.1?

**Comment by yangshengjie** [comment]
Yes. I think It would be much easier to learn it from beginning

**Comment by marpaia** [comment]
We've added a lot of features and functionality that make using osquery a lot easier since then. We definitely suggest that, if you're getting started with osquery, you try the most recent release.

**Comment by theopolis** [comment]
You'll have a much easier time with:

```
git checkout -b 1.4.4 1.4.4
make deps
```

And the user guide at: http://osquery.readthedocs.org will make a lot more sense. :)

**Comment by yangshengjie** [comment]
When I try to compile the most recent version, I got lots of errors.

**Comment by theopolis** [comment]
Are you using CentOS 6.5, CentOS 7, Ubuntu12, or Ubuntu14? Those are the only Linux distributions we support building on right now. If you're using another flavor/distro you can try using our Vagrant boxes:

Check out: http://osquery.readthedocs.org/development/building/#building-on-linux

**Comment by yangshengjie** [comment]
Ubuntu14.04

**Comment by marpaia** [comment]
try using our vagrant VMs, as the link that @theopolis posts outlines. if you'd like to use your own host `make deps` should configure your machine and install build dependencies, but we only "officially" support building on our vagrant build hosts. otherwise, you can get a deb from osquery.io/downloads/ which should work out of the box.

**Comment by yangshengjie** [comment]
/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:151:48: error: 
      too many arguments provided to function-like macro invocation
    ZVAL_STRING(&gettransport, "getTransport", 0);
                                               ^
/usr/local/php/include/php/Zend/zend_API.h:558:9: note: macro 'ZVAL_STRING'
      defined here
#define ZVAL_STRING(z, s) do {                                  \
        ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:151:5: error: 
      use of undeclared identifier 'ZVAL_STRING'
    ZVAL_STRING(&gettransport, "getTransport", 0);
............

**Comment by marpaia** [comment]
a few questions: 

- what host is this? your own host? our vagrant vm?
- can you please post a full log, wrapped in three backticks, so that the output is fixed width

**Comment by yangshengjie** [comment]
```
In file included from /usr/local/php/include/php/Zend/zend.h:31:
/usr/local/php/include/php/Zend/zend_types.h:654:9: warning: 'Z_ADDREF_P' macro
      redefined
#define Z_ADDREF_P(pz)                          zval_addref_p(pz)
        ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:61:9: note: 
      previous definition is here
#define Z_ADDREF_P ZVAL_ADDREF
        ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:151:48: error: 
      too many arguments provided to function-like macro invocation
    ZVAL_STRING(&gettransport, "getTransport", 0);
                                               ^
/usr/local/php/include/php/Zend/zend_API.h:558:9: note: macro 'ZVAL_STRING'
      defined here
#define ZVAL_STRING(z, s) do {                                  \
        ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:151:5: error: 
      use of undeclared identifier 'ZVAL_STRING'
    ZVAL_STRING(&gettransport, "getTransport", 0);
    ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:152:5: error: 
      use of undeclared identifier 'MAKE_STD_ZVAL'
    MAKE_STD_ZVAL(t);
    ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:155:5: error: 
      no matching function for call to 'call_user_function'
    call_user_function(EG(function_table), &p, &gettransport, t, 0, NUL...
    ^~~~~~~~~~~~~~~~~~
/usr/local/php/include/php/Zend/zend_API.h:457:14: note: candidate function not
      viable: no known conversion from 'zval **' (aka '_zval_struct **') to
      'zval *' (aka '_zval_struct *') for 2nd argument; remove &
ZEND_API int call_user_function(HashTable *function_table, zval *object...
             ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:159:5: error: 
      no matching function for call to '_zval_ptr_dtor'
    zval_ptr_dtor(&t);
    ^~~~~~~~~~~~~~~~~
/usr/local/php/include/php/Zend/zend_variables.h:107:33: note: expanded from
      macro 'zval_ptr_dtor'
#define zval_ptr_dtor(zval_ptr) _zval_ptr_dtor((zval_ptr) ZEND_FILE_LINE_CC)
                                ^~~~~~~~~~~~~~
/usr/local/php/include/php/Zend/zend_variables.h:97:15: note: candidate function
      not viable: no known conversion from 'zval **' (aka '_zval_struct **') to
      'zval *' (aka '_zval_struct *') for 1st argument; dereference the argument
      with *
ZEND_API void _zval_ptr_dtor(zval *zval_ptr ZEND_FILE_LINE_DC);
              ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:241:36: error: 
      too many arguments provided to function-like macro invocation
    ZVAL_STRING(&flushfn, "flush", 0);
                                   ^
/usr/local/php/include/php/Zend/zend_API.h:558:9: note: macro 'ZVAL_STRING'
      defined here
#define ZVAL_STRING(z, s) do {                                  \
        ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:248:36: error: 
      too many arguments provided to function-like macro invocation
    ZVAL_STRING(&writefn, "write", 0);
                                   ^
/usr/local/php/include/php/Zend/zend_API.h:558:9: note: macro 'ZVAL_STRING'
      defined here
#define ZVAL_STRING(z, s) do {                                  \
        ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:254:40: error: 
      too many arguments provided to function-like macro invocation
    ZVAL_STRINGL(args[0], newbuf, len, 0);
                                       ^
/usr/local/php/include/php/Zend/zend_API.h:554:9: note: macro 'ZVAL_STRINGL'
      defined here
#define ZVAL_STRINGL(z, s, l) do {                              \
        ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:241:5: error: 
      use of undeclared identifier 'ZVAL_STRING'
    ZVAL_STRING(&flushfn, "flush", 0);
    ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:243:5: error: 
      no matching function for call to 'call_user_function'
    call_user_function(EG(function_table), &t, &flushfn, &ret, 0, NULL T...
    ^~~~~~~~~~~~~~~~~~
/usr/local/php/include/php/Zend/zend_API.h:457:14: note: candidate function not
      viable: no known conversion from 'zval **' (aka '_zval_struct **') to
      'zval *' (aka '_zval_struct *') for 2nd argument; remove &
ZEND_API int call_user_function(HashTable *function_table, zval *object...
             ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:248:5: error: 
      use of undeclared identifier 'ZVAL_STRING'
    ZVAL_STRING(&writefn, "write", 0);
    ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:253:5: error: 
      use of undeclared identifier 'MAKE_STD_ZVAL'
    MAKE_STD_ZVAL(args[0]);
    ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:254:5: error: 
      use of undeclared identifier 'ZVAL_STRINGL'
    ZVAL_STRINGL(args[0], newbuf, len, 0);
    ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:258:5: error: 
      no matching function for call to 'call_user_function'
    call_user_function(EG(function_table), &t, &writefn, &ret, 1, args T...
    ^~~~~~~~~~~~~~~~~~
/usr/local/php/include/php/Zend/zend_API.h:457:14: note: candidate function not
      viable: no known conversion from 'zval **' (aka '_zval_struct **') to
      'zval *' (aka '_zval_struct *') for 2nd argument; remove &
ZEND_API int call_user_function(HashTable *function_table, zval *object...
             ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:259:5: error: 
      no matching function for call to '_zval_ptr_dtor'
    zval_ptr_dtor(args);
    ^~~~~~~~~~~~~~~~~~~
/usr/local/php/include/php/Zend/zend_variables.h:107:33: note: expanded from
      macro 'zval_ptr_dtor'
#define zval_ptr_dtor(zval_ptr) _zval_ptr_dtor((zval_ptr) ZEND_FILE_LINE_CC)
                                ^~~~~~~~~~~~~~
/usr/local/php/include/php/Zend/zend_variables.h:97:15: note: candidate function
      not viable: no known conversion from 'zval *[1]' to 'zval *'
      (aka '_zval_struct *') for 1st argument
ZEND_API void _zval_ptr_dtor(zval *zval_ptr ZEND_FILE_LINE_DC);
              ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:262:13: error: 
      cannot initialize a variable of type 'zval *' (aka '_zval_struct *') with
      an lvalue of type 'zend_object *' (aka '_zend_object *')
      zval* ex = EG(exception);
            ^    ~~~~~~~~~~~~~
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:282:42: error: 
      too many arguments provided to function-like macro invocation
      ZVAL_STRING(&putbackfn, "putBack", 0);
                                         ^
/usr/local/php/include/php/Zend/zend_API.h:558:9: note: macro 'ZVAL_STRING'
      defined here
#define ZVAL_STRING(z, s) do {                                  \
        ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:290:50: error: 
      too many arguments provided to function-like macro invocation
      ZVAL_STRINGL(args[0], newbuf, buffer_used, 0);
                                                 ^
/usr/local/php/include/php/Zend/zend_API.h:554:9: note: macro 'ZVAL_STRINGL'
      defined here
#define ZVAL_STRINGL(z, s, l) do {                              \
        ^
/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp:369:36: error: 
      too many arguments provided to function-like macro invocation
    ZVAL_STRING(&funcname, "read", 0);
                                   ^
/usr/local/php/include/php/Zend/zend_API.h:558:9: note: macro 'ZVAL_STRING'
      defined here
#define ZVAL_STRING(z, s) do {                                  \
        ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
1 warning and 20 errors generated.
make[6]: *** [php_thrift_protocol.lo] Error 1
make[6]: Leaving directory `/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php/src/ext/thrift_protocol'
make[5]: *** [src/ext/thrift_protocol/modules/thrift_protocol.so] Error 2
make[5]: Leaving directory `/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php'
make[4]: *** [all-recursive] Error 1
make[4]: Leaving directory `/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib/php'
make[3]: *** [all-recursive] Error 1
make[3]: Leaving directory `/home/sunan/opensource/osquery/.sources/thrift-0.9.1/lib'
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory `/home/sunan/opensource/osquery/.sources/thrift-0.9.1'
make[1]: *** [all] Error 2
make[1]: Leaving directory `/home/sunan/opensource/osquery/.sources/thrift-0.9.1'
make: *** [deps] Error 2
```
On my own host. 
BTW, you guy are very nice, thanks very much!

**Comment by theopolis** [comment]
@yangshengjie, I've never seen those errors before, but you can try:

```
cd /home/sunan/opensource/osquery/.sources/thrift-0.9.1/
./configure --with-cpp=yes --with-ruby=no --with-go=no --with-erlang=no --with-java=no --with-python=no --with-php=no
make
sudo make install
cd /home/sunan/opensource/osquery/
make deps
make
```

**Comment by yangshengjie** [comment]
@theopolis I have another question, I found there is a third-party and .source folder under this project. what is the difference between those tow folders?

**Comment by marpaia** [comment]
third-party is where we keep the code that we build inline as apart of the osquery build process. sources is where we download sources that we need to augment build hosts. it's a subtle, but real difference.

**Comment by theopolis** [comment]
Right, we maintain some patches and modifications to a few of our dependencies. Those modified sources are kept in third-party.

`make deps` uses a "provisioning" script/helper to turn the target OS into a development machine. Part of the provisioning process includes pulling down a few unmodified dependency sources for software versions not included in the distribution's package manager. Those go into the temporary/working folder ".sources".

**Comment by theopolis** [comment]
@Red-wolf-s-husband, did that small change to how we configure thrift help? If so I can make that change to our [install script](https://github.com/facebook/osquery/blob/master/tools/provision/lib.sh#L41).

**Comment by Red-wolf-s-husband** [comment]
Yes, problem solved, I just wake up and made a pull request. But found you already made one. Can you adapt mine?

**Comment by marpaia** [comment]
Merged in @theopolis' (#978)
